### ISSUE TITLE
Implement the revised Records displays for FreeCEN2

### ISSUE THREAD

**Issue by Captainkirkdawson** [FreeUKGen/FreeCENMigration#1265]
A phased implementation plan is as follows:
a. Implement changes to Freecen1 and Freecen2 so that all pieces have a process to set status field to 'Online' and populate an Online_date field. (see #1202 for freecen2 - need new story for freecen1?)
b. Develop new Records pages (excluding Parish data, unique names and charts).
c. Develop unique names capability at Piece level.
d. Develop charts.
e. Investigate capabilities for providing unique names at higher levels (performance testing on Test3?) and implement where possible.
f. Records data integration at Parish level for VLD and CSVProc data.
[https://app.zenhub.com/files/28748917/25dae190-0141-492c-a093-07572fb8d49c/download](https://app.zenhub.com/files/28748917/25dae190-0141-492c-a093-07572fb8d49c/download)

**Comment by DeniseColbert** [comment]
Looking at compatibility between pieces in FC1 & FC2

**Comment by Captainkirkdawson** [comment]
@AnneV-Learn has written a rake task to determine the extent to which freecen pieces can be linked to Freecen2Places. With a series of iterations we were able to achieve 99% correct connections. We have added some additional alternate place names that should bring us to 100%. 
**@PatReynolds @geoffj-FUG The Freecen2Place connections currently are set to follow the same place pattern as FC1 which is at Sub District, District and occasional Civil Parish level; we could change to a consistent District level if desired.**
Next step is to write a rake task to populated the freecen_piece document with a freecen2_place_id

**Comment by geoffj-FUG** [comment]
Kirk

 

What are the advantages of consistency vs the effort involved.

 

For instance if the tables are going to be used for searching then as much
consistency as possible would be advantageous to make the production of
search results easier. 

 

Would changing to a consistent District level affect the recent PARMS
geolocating exercise? If so I don't think there is the goodwill out there at
the moment. We need to have a period of stability to rebuild confidence in
the systems.

 

Geoff

**Comment by Captainkirkdawson** [comment]
@geoffj-FUG I believe it would perhaps lead to some confusion. Lets then go with the pragmatic approach at this time until  there is enough desire to change.

**Comment by Captainkirkdawson** [comment]
@AnneV-Learn I believe you can now proceed with the development of the records displays based on FreeCEN2Place.

**Comment by DeniseColbert** [comment]
Kirk to spin up test 2 for Pat and Denise to test when ready

**Comment by DeniseColbert** [comment]
Coming along nicely, Anne will tune it for optimisation ready for the next steps.

**Comment by PatReynolds** [comment]
Coud the status of a piece be more informative - i.e. 'online' 'in progress' 'pending' or 'unavailable' - the last with a note in the note field explaining that the piece was lost overboard en route to gurnsey, burnt, or whatever.

**Comment by AnneV-Learn** [comment]
@PatReynolds I think this may have to be raised as a separate story so that new functionality to set the status can be built into the freecen2 Piece management processes. Having completed that any new status settings could be reflected in the Records pages. It is my understanding that we currently have Statuses of Online, Partial(used when one or more CSV files have be incorporated for a piece but not all) and ‘blank’ when no data has been incorporated for the Piece yet.  
Do you agree @Captainkirkdawson?

**Comment by Captainkirkdawson** [comment]
I agree completely @AnneV-Learn this is a brand new requirement. 
@PatReynolds the status field is a strange field dating back to the early days of FC and was used to indicate the status of a vld file for a piece during it loading. It only ever had 4 possible values; 'Online', 'Loading',  'Error (not unique)' and ''. CSVProc added 'Part'
The only 2 values that actually exist within freecen_piece are 'Online' and '' as the 2 error conditions have all been resolved.. For freecen2_piece 'Online', 'Part' and '' are in use.
Assigning new values would likely be part of an online work flow management process. IMO freecen management is no where ready to define and such a process.

**Comment by PatReynolds** [comment]
Ready for testing.  @DeniseColbert  to ask Chris to test once we have done a little more testing ourselves.

**Comment by Captainkirkdawson** [comment]
Could I suggest we add an extra link and a new column. The link is a location line that opens a map with the location of the place. The new column lists the number of records on line.
See below
![Place display.png](https://images.zenhubusercontent.com/5b68ae684b5806bc2bdefcf0/c16ebb53-1f91-46cb-a61e-f8fe68bb3bda)![Place display.png]
![pieces display.png](https://images.zenhubusercontent.com/5b68ae684b5806bc2bdefcf0/e6be59f7-a363-4c79-90bb-c303aaa4a3a9)

**Comment by geoffj-FUG** [comment]
I support both of Kirk's suggestions. Geoff

**Comment by AnneV-Learn** [comment]
@Captainkirkdawson I am wondering if it would also be useful to add a button on the 'All Counties' page that links to a list of Places with New Pieces Online  i.e. added in previous 30 days. The list would show County, Place, Year and buttons to 'View Pieces' and 'Unique Names').

**Comment by Captainkirkdawson** [comment]
Yes  it would. It would also likely be a good idea to add it to the county page too

**Comment by AlOneill** [comment]
@AnneV-Learn I've just had a quick look at the new Records page — it is looking good!

Are you ready for me to hit you with some accessibility issues? Or should I hold off for now?

**Comment by AnneV-Learn** [comment]
@AlOneill please do send me the accessibility issues.

**Comment by AlOneill** [comment]
@AnneV-Learn Starting with the Records page —

* although using a `fieldset` to hold both the form and the table is not invalid markup per se, I think it would confuse screen-reader software and hence those who use that software. The `fieldset` could contain the whole form, in which case it would need to have a `form="id-of-form"` attribute. It is more commonly a child of a form.

Further, a screen-reader user might have to hear the `legend` announced for every input, so brevity is good.

* the form `input` elements should have a `label` such that the `for` attribute of the label is the same as the `id` attribute of the input it belongs to: for example, `<label for="county_description" … >Select a county</label>`

* the `li` elements must be contained in a `ul` or an `ol`

* "County" and "Place and Unique Names" look like headings but are not marked up as such

* I'm assuming that the link to the original Records page is a temporary one? If not, 'HERE' is not acceptable as the link text, but you knew that :-)

* there is a stray `<em>`, at the bottom of the content which is having an impact on the 'Top of page' button (when logged in)

* the form needs a `caption` as its first child: this explains the structure of the form to a non-sighted user and so could be hidden from sighted users with `class="accessibility"`

It would be good not to use `<br>` to control spacing. It is fine for breaking a line, as in the table headings, but not to push an element down the page. If the spacing issue you have now is not resolved by fixing the other issues, then we can reassess and apply a class to fix things.

These issues are mostly ones that I've noticed myself. I cannot use the WebAim accessibility checker as the page is behind a login. If you can arrange things to check the page (which must be online) using the [Wave accessibility tool](https://wave.webaim.org/) yourself, then so much the better.

Do get back to me if any of he above does not make sense!

**Comment by AnneV-Learn** [comment]
@AlOneill  thank you for your preliminary review. By ‘behind a login’ do you mean the Test3 login (just checking as Records pages should be available to non-members)?
I am currently working on some additional functionality and so I think I will deal with that first now.  
To be honest my html experience is very minimal and I just aimed to get the Records landing page (which is the most complicated of the Records pages) to look ‘tidy’. I will most certainly need some help to resolve accessibility issues.

**Comment by AlOneill** [comment]
@AnneV-Learn Yes, the t3 login. There won't be an issue once the page is on production (!)

No problem with your lack of experience: we all have to learn something new with these projects.

**Comment by AnneV-Learn** [comment]
@AlOneill I have addressed the issues you identified and also checked all the pages in the RECORDS functionality with the WAVE accessibility tool and made corrections where necessary. Now none of the pages have Errors or Alerts associated with the RECORDS views. So hopefully all will be ok when the updated version is made available for testing.

**Comment by PatReynolds** [comment]
I can't see the Unique names button (on test 3). Should I?

The selection of places does not work ain the same way as on the search page - typing in a character will take you to the first instance of  a place starting with the character.  Is it possible to add this, for consistency and ease of use.

Otherwise all working as expected, thank you!

**Comment by AnneV-Learn** [comment]
The Unique names button only appears on the Place index if there is data Online. I don’t know which counties/places you tried @PatReynolds but if you try County = Lancashire and then Place = Accrington hopefully you will see the unique names button for 1861 data. 
Tomorrow I will be releasing an updated version of the Records  pages for preliminary testing by Kirk. This includes new functionality (see previous suggestions in the thread) and will also address your second issue (the type of drop down I had originally used for the Places dropdown on the County totals page did not meet the accessibility criteria so it has been changed).

**Comment by AlOneill** [comment]
@AnneV-Learn Excellent — thank you! I look forward to being able to re-test all the Records pages.

**Comment by Captainkirkdawson** [comment]
We have pulled the code into test3 but the re appear to be underlying data issues so taking the time to redo the linkages. The displays look excellent; may take a day or 2 to get the database setup on test3

**Comment by PatReynolds** [comment]
Unique names button now there - I must have not been looking in the right place.  Everything works as I would expect it to (and very easy to use!)

Is there a typo in Stockton 1851, piece HO107_2383A  "Norton, (Blakeston), Stainton(Thornaby), Stockton upon Tees(East Hartburn;Preston-upon-Tees;Stockton-upon Tees) " - missing second hyphen in last name.

**Comment by Captainkirkdawson** [comment]
@PatReynolds the spelling issue is nothing to do with the Records display; it is an incorrect name in the TNA hamlet.

**Comment by Captainkirkdawson** [comment]
We have identified 2 issues.
1) VLD file deletions were unclean in that they left some fields uncleared.  These can only be dealt with AFTER deployment to production on a case by case basis.
2) CSVProc uses the Civil Parish place not the Piece place and a piece may have multiple cavil parishes

**Comment by PatReynolds** [comment]
1) @AnneV-Learn to add code so system admin access new code  - everyone else to see current version - for clean up of (a).

2) Decision to go for granularity ie use Civil Parish Place. Deploy after (1) completed, and and work on (2).

**Comment by AlOneill** [comment]
@AnneV-Learn 
On the Records page, there is a `div.grid__item` as the first child of the `div.container`: it is a quirk of our CSS that all `.grid__item` elements must be children of a `.grid` element (and vice-versa). If not, the spacing gets messed up, especially in a narrow window.

![records-01Dec.png](https://images.zenhubusercontent.com/5ce6b34165b1a70acc0a9124/080f9c48-0637-46b7-98a6-5881371cb3eb)

I think there could be a better way of letting a screen-reader user understand the structure of the page: perhaps by making the `h1.accessibility` a span that is a child of the main page heading?

The labels are now associated with their inputs — thank you.

A `label.ttip` is not accessible to a sighted keyboard-only user _unless_ it has the attribute `tabindex="0"` (they cannot hover).

**Comment by AlOneill** [comment]
@AnneV-Learn The use of brackets to convey meaning is likely lost on screen-reader users as punctuation is not announced typically.

There is also the related issue of counties listed as "Angus (Forfarshire)", for example, because two names were in use — what will a user make of the brackets used there? That Angus has records but that Forfarshire does not?? Those in the know will probably understand, but what about the remainder? Maybe "Angus or Forfarshire" would be clearer all round?

**Comment by AnneV-Learn** [comment]
@AlOneill  - I have attempted to address your observations  and the revisions should be available to review when the next version is moved to Test3. I'll let you know when that happens.

**Comment by Captainkirkdawson** [comment]
@PatReynolds I have not proceeded with deployment to production as I have found no way of merging the civil parishes place data with the piece place data. None of the strategies I have tried work. Since this means that CSV loads would not be reported  an alternative solution should be found prior to deployment. Moving back to in progress.

**Comment by AnneV-Learn** [comment]
@PatReynolds @AlOneill I actually found a solution which appears to resolve the issue of merging the civil parish place and piece place data.  The rake task to collect the data does take longer to run now but I’ll continue to investigate possible fixes for that. 
I also changed the County and Place drop-downs  to improve accessibility. The Place drop-down now includes the census ‘years with data’ in brackets (follows the look of the Census Place drop-down on the main FreeCen Search page). 
Kirk has moved to Test3 and is moving to Production, both  with a switch that means these new Records pages are currently only visible to System Administrators so that we can review with the production data.
